### PR TITLE
Added connection pool metrics

### LINKED ISSUE #521: Connection Pool Metrics
## Detailed Description

The http client module could include support for metrics on the underlying connection pool, e.g. number of connections per state.

## Context

Investigating issues with the connection pool are tricky right now because there is virtually no insight into the connection pool.

## Possible Implementation

To be defined.

## Your Environment

* Version used: 2.10.1
* Link to your project:

### PR DESCRIPTION
**Pull request by whiskeysierra**
## Motivation and Context

Fixes #521 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

### CODE REVIEW

File: riptide-httpclient/src/main/java/org/zalando/riptide/httpclient/metrics/HttpConnectionPoolMetrics.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    @Override
+    public void bindTo(final MeterRegistry registry) {
+        gauge(registry, "available", PoolStats::getAvailable);
+        gauge(registry, "leased", PoolStats::getLeased);
+        gauge(registry, "max", PoolStats::getMax);
+        gauge(registry, "pending", PoolStats::getPending);
+    }
+
+    private void gauge(final MeterRegistry registry, final String name, final Function<PoolStats, Number> function) {
+        Gauge.builder(metricName + "." + name, () -> function.apply(manager.getTotalStats()))
```
Comment (lukasniemeier-zalando) [inline riptide-httpclient/src/main/java/org/zalando/riptide/httpclient/metrics/HttpConnectionPoolMetrics.java]: `getTotalStats` potentially locks the connection pool - do we have to protect people from sampling this `Guage` too often (assuming micrometer does not cache values anyhow)?

Comment (whiskeysierra) [reply riptide-httpclient/src/main/java/org/zalando/riptide/httpclient/metrics/HttpConnectionPoolMetrics.java]: I'd say it's up to the monitoring system and the choice of our users how often metrics are published/sampled.

File: riptide-httpclient/src/main/java/org/zalando/riptide/httpclient/metrics/HttpConnectionPoolMetrics.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    public HttpConnectionPoolMetrics withDefaultTags(final Iterable<Tag> defaultTags) {
+        return new HttpConnectionPoolMetrics(manager, metricName, ImmutableList.copyOf(defaultTags));
+    }
+
+    @Override
+    public void bindTo(final MeterRegistry registry) {
+        // since getTotalStats locks the connection pool, we cache the value for a minute to reduce possible contention
+        final Supplier<PoolStats> stats = memoizeWithExpiration(manager::getTotalStats, 1, MINUTES);
```
Comment (whiskeysierra) [inline riptide-httpclient/src/main/java/org/zalando/riptide/httpclient/metrics/HttpConnectionPoolMetrics.java]: @lukasniemeier-zalando I implemented caching now as you suggested. Is 1 minute a good value? Should we make this configurable?

Comment (lukasniemeier-zalando) [reply riptide-httpclient/src/main/java/org/zalando/riptide/httpclient/metrics/HttpConnectionPoolMetrics.java]: +1 on going with 1 minute as a default. let's expand this as soon as the use case of customization pops up.

**Comment by lukasniemeier-zalando** [comment]
:+1:

**Pull request merged**